File: flux-tensor-midi/flux_tensor_midi/core/clock.py
"""T-0 clock with EWMA adaptation.

The T-0 clock is an agent's temporal expectation engine.
It tracks when the next "tick" should happen, adapts its interval
using exponentially weighted moving averages, and reports drift.
"""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class TZeroClock:
    """Agent's temporal expectation clock with EWMA adaptation.

    The clock tracks an expected interval and learns from actual
    observations using exponential moving average.

    Attributes:
        interval: Expected interval in seconds.
        adaptive: Whether to use EWMA to adjust the interval.
    """

    interval: float = 300.0
    adaptive: bool = True

    def __post_init__(self) -> None:
        self._t_last: float = time.time()
        self._t_zero: float = self._t_last + self.interval
        self._ewma_alpha: float = 0.1
        self._tick_count: int = 0

    @property
    def t_zero(self) -> float:
        """When the next tick is expected."""
        return self._t_zero

    @property
    def t_last(self) -> float:
        """When the last tick occurred."""
        return self._t_last

    @property
    def tick_count(self) -> int:
        """Total ticks since creation."""
        return self._tick_count

    @property
    def bpm(self) -> float:
        """Convert current interval to BPM (beats per minute).

        A 300s interval = 0.2 BPM (very slow).
        A 1s interval = 60 BPM.
        A 0.5s interval = 120 BPM.
        """
        if self.interval <= 0:
            return 0.0
        return 60.0 / self.interval

    @bpm.setter
    def bpm(self, value: float) -> None:
        """Set interval from BPM value."""
        if value > 0:
            self.interval = 60.0 / value
            self._recalculate_t_zero()

    def _recalculate_t_zero(self) -> None:
        """Recalculate t_zero from t_last and current interval."""
        self._t_zero = self._t_last + self.interval

    def tick(self) -> Optional[float]:
        """Check if T-0 has passed.

        Returns:
            Delta (actual - expected) if T-0 passed, None if on time.
            Positive delta = late, negative delta = early.
        """
        now = time.time()
        if now >= self._t_zero:
            actual_interval = now - self._t_last
            delta = actual_interval - self.interval
            self._t_last = now
            self._tick_count += 1

            if self.adaptive:
                self.observe(actual_interval)

            self._recalculate_t_zero()
            return delta
        return None

    def observe(self, actual_interval: float) -> None:
        """Learn from actual interval using EWMA.

        Args:
            actual_interval: The observed interval in seconds.
        """
        self.interval = (
            self._ewma_alpha * actual_interval
            + (1.0 - self._ewma_alpha) * self.interval
        )
        self.interval = max(0.01, self.interval)  # Floor at 10ms

    def missed_ticks(self, elapsed: Optional[float] = None) -> int:
        """How many expected ticks were missed in the elapsed time.

        Args:
            elapsed: Time elapsed. Defaults to time since last tick.

        Returns:
            Number of missed ticks (0 if none missed).
        """
        if elapsed is None:
            elapsed = time.time() - self._t_last
        if self.interval <= 0:
            return 0
        expected = elapsed / self.interval
        actual = elapsed / (elapsed + 0.001)
        return max(0, int(expected) - 1)

    def reset(self, interval: Optional[float] = None) -> None:
        """Reset the clock.

        Args:
            interval: New interval. Defaults to current.
        """
        if interval is not None:
            self.interval = interval
        self._t_last = time.time()
        self._recalculate_t_zero()

    def remaining(self) -> float:
        """Seconds until next T-0."""
        return max(0.0, self._t_zero - time.time())

    def __repr__(self) -> str:
        return (
            f"TZeroClock(interval={self.interval:.2f}s, "
            f"bpm={self.bpm:.1f}, ticks={self._tick_count})"
        )

File: flux-tensor-midi/rust/src/core/clock.rs
//! T-0 temporal expectation clock with EWMA adaptation.
//!
//! Tracks when the next observation is expected (`t_zero`) and adapts the interval
//! using exponential weighted moving average (EWMA).

use core::fmt;

#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

/// T-0 temporal expectation clock.
///
/// Learns the interval between observations and predicts the next one.
/// Uses EWMA to adapt to tempo changes.
#[derive(Clone, Debug)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct TZeroClock {
    /// Expected interval between observations (seconds).
    pub interval: f64,
    /// Timestamp of the last observation.
    pub t_last: f64,
    /// Predicted time of next observation.
    pub t_zero: f64,
    /// EWMA adaptation rate `[0, 1]`. Higher = faster adaptation.
    pub ewma_alpha: f64,
    /// Whether the clock has received at least one observation.
    pub adaptive: bool,
    /// Total observations received.
    pub observations: u64,
}

impl TZeroClock {
    /// Create a new clock with an initial interval estimate.
    ///
    /// `initial_interval` is in seconds (e.g., 0.5 for 120 BPM quarter notes).
    pub fn new(initial_interval: f64) -> Self {
        Self {
            interval: initial_interval,
            t_last: 0.0,
            t_zero: initial_interval,
            ewma_alpha: 0.3,
            adaptive: false,
            observations: 0,
        }
    }

    /// Create a clock from a BPM value.
    ///
    /// `subdivision` is the note value (4 = quarter note, 8 = eighth note).
    pub fn from_bpm(bpm: f64, subdivision: u32) -> Self {
        let interval = 60.0 / bpm / (subdivision as f64 / 4.0);
        Self::new(interval)
    }

    /// Observe an event at `t_now`. Updates interval via EWMA and predicts next.
    ///
    /// Returns the deviation (actual interval - expected interval).
    pub fn observe(&mut self, t_now: f64) -> f64 {
        if !self.adaptive {
            self.t_last = t_now;
            self.t_zero = t_now + self.interval;
            self.adaptive = true;
            self.observations = 1;
            return 0.0;
        }

        let observed_interval = t_now - self.t_last;
        let deviation = observed_interval - self.interval;

        // EWMA update
        self.interval = self.ewma_alpha * observed_interval + (1.0 - self.ewma_alpha) * self.interval;
        self.t_last = t_now;
        self.t_zero = t_now + self.interval;
        self.observations += 1;

        deviation
    }

    /// How far we are from the predicted T-0, in seconds.
    ///
    /// Negative = early, positive = late, 0 = right on time.
    #[inline]
    pub fn drift(&self, t_now: f64) -> f64 {
        if !self.adaptive {
            return 0.0;
        }
        t_now - self.t_zero
    }

    /// Normalized drift as a fraction of the expected interval.
    ///
    /// Returns 0.0 if no interval is set.
    pub fn drift_normalized(&self, t_now: f64) -> f64 {
        if !self.adaptive || self.interval == 0.0 {
            return 0.0;
        }
        self.drift(t_now) / self.interval
    }

    /// Whether `t_now` is within `tolerance` seconds of the expected T-0.
    #[inline]
    pub fn on_time(&self, t_now: f64, tolerance: f64) -> bool {
        self.drift(t_now).abs() <= tolerance
    }

    /// Reset the clock, keeping the learned interval.
    pub fn reset(&mut self) {
        self.t_last = 0.0;
        self.t_zero = self.interval;
        self.adaptive = false;
        self.observations = 0;
    }

    /// Set the EWMA alpha (adaptation rate).
    pub fn set_alpha(&mut self, alpha: f64) {
        self.ewma_alpha = alpha.clamp(0.0, 1.0);
    }

    /// Expected BPM based on current interval.
    pub fn bpm(&self) -> f64 {
        if self.interval > 0.0 { 60.0 / self.interval } else { 0.0 }
    }
}

impl fmt::Display for TZeroClock {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "TZero(interval={:.3}s, bpm={:.1}, obs={})",
            self.interval,
            self.bpm(),
            self.observations
        )
    }
}
